src/quic/streams.cc

```diff
@@ -1,61 +1,1052 @@
+    }
+
+    if (ret == bob::Status::STATUS_BLOCK) {
+      // If the read returns blocked, and there are uncommitted bytes in the
+      // queue, we'll return the current set of uncommitted bytes.
+      if (uncommitted_ > 0) {
+        PullUncommitted(std::move(next));
+        return bob::Status::STATUS_CONTINUE;
+      }
+      // If the read returns blocked, and there are no uncommitted bytes in the
+      // queue, we'll return blocked.
+      std::move(next)(bob::Status::STATUS_BLOCK, nullptr, 0, [](int) {});
+      return bob::Status::STATUS_BLOCK;
+    }
+
+    // Reads here are generally expected to be synchronous. If we have a reader
+    // that insists on providing data asynchronously, then we'll have to block
+    // until the data is actually available.
+    if (ret == bob::Status::STATUS_WAIT) {
+      next_pending_ = true;
+      std::move(next)(bob::Status::STATUS_BLOCK, nullptr, 0, [](int) {});
+      return bob::Status::STATUS_BLOCK;
+    }
+
+    CHECK_EQ(ret, bob::Status::STATUS_CONTINUE);
+    PullUncommitted(std::move(next));
+    return bob::Status::STATUS_CONTINUE;
+  }
+
+  void MemoryInfo(MemoryTracker* tracker) const override {
+    tracker->TrackField("queue", queue_);
+    tracker->TrackField("reader", reader_);
+    tracker->TrackFieldWithSize("buffer", total_);
+  }
+
+  SET_MEMORY_INFO_NAME(Stream::Outbound)
+  SET_SELF_SIZE(Outbound)
+
+ private:
+  struct OnComplete {
+    bob::Done done;
+    explicit OnComplete(bob::Done done) : done(std::move(done)) {}
+    ~OnComplete() { std::move(done)(0); }
+  };
+
+  void PullUncommitted(bob::Next<ngtcp2_vec> next) {
+    MaybeStackBuffer<ngtcp2_vec, 16> chunks;
+    chunks.AllocateSufficientStorage(count_);
+    auto head = commit_head_;
+    size_t n = 0;
+    while (head != nullptr && n < count_) {
+      // There might only be one byte here but there should never be zero.
+      DCHECK_LT(head->offset, head->buf.len);
+      chunks[n].base = head->buf.base + head->offset;
+      chunks[n].len = head->buf.len - head->offset;
+      head = head->next.get();
+      n++;
+    }
+    std::move(next)(bob::Status::STATUS_CONTINUE, chunks.out(), n, [](int) {});
+  }
+
+  void MarkErrored() {
+    errored_ = true;
+    head_.reset();
+    tail_ = nullptr;
+    commit_head_ = nullptr;
+    total_ = 0;
+    count_ = 0;
+    uncommitted_ = 0;
+    MarkEnded();
+  }
+
+  void MarkEnded() {
+    eos_ = true;
+    queue_.reset();
+    reader_.reset();
+  }
+
+  void Append(const DataQueue::Vec* vectors, size_t count, bob::Done done) {
+    if (count == 0) return;
+    // The done callback should only be invoked after we're done with
+    // all of the vectors passed in this call. To ensure of that, we
+    // wrap it with a shared pointer that calls done when the final
+    // instance is dropped.
+    auto on_complete = std::make_shared<OnComplete>(std::move(done));
+    for (size_t n = 0; n < count; n++) {
+      if (vectors[n].len == 0 || vectors[n].base == nullptr) continue;
```

Comment:
Is performance critical for this function? There are too many branches in this for loop.

Reply:
Performance is important here, yes, but for now I want to focus on the core functionality. Make it work first, then make it fast. I'll make a note to revisit this.